### PR TITLE
feature/ND/230/Node-Event-Scheduler

### PR DESCRIPTION
**Pull request by ShaneTWilliams**
Welcome to ghetto RTOS, folks

`EventScheduler` is responsible for calling callback functions at given intervals, essentially creating a system of (many) non-blocking delays. It's very heavily based on the Arduino timer we used last year.

The `EventScheduler` class owns an array of 50 `Event` objects, each of which has a delay, a count of how many times it needs to be called, a type, a state, the last time it was called, and a callback function. Through the `EventScheduler` class, you can set one of these `Event`s to run a callback every X seconds. 50 is an arbitrary number which can be changed in the future.

`EventScheduler` has 3 methods that matter:
- `EventScheduler::callFunctionAfter` calls the callback once after X seconds
- `EventScheduler::callFunctionEvery` (version 1) calls the callback every X seconds
- `EventScheduler::callFunctionEvery` (version 2) calls the callback every X seconds, but only up until the specified number of calls has been reached

Each of these returns an `eventId_t ` ID which is used to identify that event, and can be handy when using the class's `startEvent` and `stopEvent` methods. Also, whenever you call one of those 3 methods, the class automatically assigns you an unused event object, and therefore an unused event ID.

Every loop in the main, you **_NEED_** to call the `EventScheduler::updateEvents` function. This is what checks every event to see if its delay has passed. If so, the event's callback is called. If the event has been run its maximum number of times, its `Event::cleanup` function is called, and it resets all its properties and becomes ready to be assigned again.

There's also unit tests under `OnPod/Node/Test/EventScheduler`. Pls review them too.

P.S. - the `Event` class definitely could have been been made a struct, but the current solution is super clean and leaves room for the future if needed.

### CODE REVIEW

File: OnPod/Node/Common/EventScheduler/include/event.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        void setLastTimeCalledMs(timeMs_t time_ms);
+        void setCallback(void (*callback)(void));
+        void setNumCallsLeft(uint32_t num_calls);
+
+        // Take away one from the event's `m_num_calls_left`
+        void decrementNumCallsLeft(void);
+        void callCallback(void);
+        // Reset event attributes to prepare for reassignment
+        void cleanup(void);
+};
```
Comment (colton-smith) [inline OnPod/Node/Common/EventScheduler/include/event.h]: This is really clean - but I wonder if making the `EventScheduler` class a friend of the `Event` class would be a cleaner way of doing it. Looks as though the `Event` class is only used by the `EventScheduler` class, not anywhere else. Using a friend class would replace all the getters and setters by providing the `EventScheduler` class access to the `Event` classes private members.  Might be less robust that way though. Let me know what you think.

File: OnPod/Node/Test/EventScheduler/main.cpp [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+void stopEvent_eventStopped_eventStaysStopped(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionAfter(1000, testCallback, 0);
+    event_scheduler.stopEvent(event);
+    event_scheduler.stopEvent(event);
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(0, test_callback_run_count);
+    resetTestCallback();
+}
```
Comment (colton-smith) [inline OnPod/Node/Test/EventScheduler/main.cpp:50]: `stopEvent` occurs twice consecutively here on lines 45 and 46, intentional?

File: OnPod/Node/Test/EventScheduler/main.cpp [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+void startEvent_eventNotStarted_eventStarted(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionAfter(1000, testCallback, 0);
+    event_scheduler.stopEvent(event);
+    event_scheduler.startEvent(event);
+    event_scheduler.updateEvents(1001);
+    TEST_ASSERT_EQUAL_UINT16(1, test_callback_run_count);
+}
+
```
Comment (colton-smith) [inline OnPod/Node/Test/EventScheduler/main.cpp]: Do you ever test the case where there are no available event_ids? The case where the `findFreeEvent` methods returns -1?

File: OnPod/Node/Test/EventScheduler/main.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+void callFunctionEveryFinite_delayPassedThreeTimes_functionCalledTwice(void)
+{
+    EventScheduler event_scheduler;
+    eventId_t event = event_scheduler.callFunctionEvery(1000, testCallback, 0, 2);
+    event_scheduler.updateEvents(1001);
+    event_scheduler.updateEvents(2002);
+    event_scheduler.updateEvents(3003);
+    TEST_ASSERT_EQUAL_UINT16(2, test_callback_run_count);
+    resetTestCallback();
+}
```
Comment (colton-smith) [inline OnPod/Node/Test/EventScheduler/main.cpp:161]: Maybe repeat this test except create a new event and ensure the event_id is 0 and not 1? To make sure the old complete event is properly cleared for use. Just to check off all the boxes?

**Review by colton-smith — COMMENTED** [review]
Loved it.
Style-wise it's very clean, no complaints here.

Although I don't have much experience with event-driven code, I'm able to understand this which is a good sign in terms of readability. Code-wise the only thing I saw is the possibility of making the `EventScheduler` class a friend of the `Event` class, as I mentioned in the comment. 

Testing is thorough - love to see it. I added a few ideas for test cases, the more boxes we check the better. 

Great job fearless leader

File: OnPod/Node/Common/EventScheduler/include/event.h (hunk 1 above)
Comment (colton-smith) [reply OnPod/Node/Common/EventScheduler/include/event.h]: On second thought, that would detract from encapsulation and also require too much unnecessary work.

File: OnPod/Node/Common/EventScheduler/src/event.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+void Event::decrementNumCallsLeft(void)
+{
+    m_num_calls_left--;
+}
+
+void Event::callCallback(void)
+{
+    m_callback();
+}
```
Comment (andrewnash) [inline OnPod/Node/Common/EventScheduler/src/event.cpp]: Basically, everything in this file is just one line returns. Would be much easier and cleaner to inline this in the header file

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                    m_events[id].cleanup();
+                }
+            }
+        }
+    }
+}
+
+eventId_t EventScheduler::findFreeEvent(void)
+{
+    for (eventId_t id = 0; id < m_MAX_NUM_EVENTS; id++)
```
Comment (andrewnash) [inline OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: It would literally be faster to check cells randomly than to linear search every time in practice. Any time you have to linear search, you should be rethinking your data structure to make sure it's the best choice, consider using a hash table or using another data structure to keep track of free event ids.

**Review by andrewnash — CHANGES_REQUESTED** [review]
These comments are an amalgamation of me and Mark briefly looking at this together. Maybe we are missing something but unsure how this is non-blocking, pretty sure would need multiple threads to accomplish this? Should these events have some priority based on the time elapsed? Maybe some form of priority queue is better suited?

File: OnPod/Node/Common/EventScheduler/src/event.cpp (hunk 5 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event.cpp]: I thought that was bad practice?

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: I figured that iterating through 50 items then making one comparison isn't really that time consuming so I didn't bother but you're definitely right

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: > Maybe we are missing something but unsure how this is non-blocking, pretty sure would need multiple threads to accomplish this? Should these events have some priority based on the time elapsed? Maybe some form of priority queue is better suited?

This code isn't blocking, it only checks for events due when you call the eventUpdate() function. I don't think that priorities are really worth the hassle

File: OnPod/Node/Test/EventScheduler/main.cpp (hunk 3 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Test/EventScheduler/main.cpp]: Yeah I'll add this. Good catch

File: OnPod/Node/Test/EventScheduler/main.cpp (hunk 4 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Test/EventScheduler/main.cpp:161]: I'm kinda confused about what you mean here

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Non-blocking refers to code that doesn't block the execution of other code. In this case, calling `callCallback` one "event" blocks calling `callCallback` of all other events until complete.

Also as a side note, after writing that sentence, I am triggered your method naming  `callCallback`.

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Yes obviously these function calls are going to block others, the plan was never to have threading on the nodes unless we really needed it. I moreso meant that this class replaces blocking `delay()` calls in our main.

Also yeah lmao at that name but it is describing exactly what the method does

Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: the word call is redundant, by calling a method it is implied that you are... calling it. I understand that you want to represent that the method is a wrapper but I don't think that info is necessary in the name

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: So I name the method `callback()`? That's the name of the attribute that holds the pointer to the callback function. What I have should be fine

File: OnPod/Node/Common/EventScheduler/src/event.cpp (hunk 5 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event.cpp]: If it's not a big deal then I'd rather change it to what you suggested cause it definitely is cleaner

Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event.cpp]: it's not bad practice when the implementation is very straight forward (one-line command) like this

File: OnPod/Node/Test/EventScheduler/main.cpp (hunk 4 above)
Comment (colton-smith) [reply OnPod/Node/Test/EventScheduler/main.cpp:161]: I just mean maybe test that finite events get cleaned up after they "expire", could be too explicit of a test though.

**Review by Burke-Daniel — COMMENTED** [review]
Looks great!

Style is clean and consistent, and I am a fan of the test function names, it makes it very explicit and easy to read.

Like Colton I don't have much experience with event driven code, however I understand what this code accomplishes, and anything I noticed has already been addessed by other's comments.

Great job!

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: `callCallback` calls `m_callback`, don't see anywhere where the name `callback` is taken?

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: I'll change it

File: OnPod/Node/Common/EventScheduler/include/event.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+#endif
```
Comment (Newfloridadad) [inline OnPod/Node/Common/EventScheduler/include/event.h]: very minor, gits picking up your lack of newline at eof

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Regarding the original comment about linear searching, are you able to use `std::unordered_map` or something similar on the node? I assume you can't use vectors, otherwise, you would have already.

File: OnPod/Node/Test/EventScheduler/main.cpp (hunk 3 above)
Comment (Newfloridadad) [reply OnPod/Node/Test/EventScheduler/main.cpp]: If the -1 is returned I would send a can message/ set a flag in a node health message if we have one to indicate the event buffer is full. Or alternatively, use a vector if it works on the hardware for flexibility overall but I am fine with the array implementation with a max capacity check.

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: I'm 99% sure I could but dynamic memory is sorta taboo with embedded applications and I've been trying to avoid it

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: @stuckatmarine any insight?

File: OnPod/Node/Test/EventScheduler/main.cpp (hunk 4 above)
Comment (Newfloridadad) [reply OnPod/Node/Test/EventScheduler/main.cpp:161]: store the id returned by event 1000, let it expire. Create event 2000 and the id should be the same as 1000, both being 0. this ensures 1000 was cleaned up and replaced properly.

Additionally, I would fill with 51 events to ensure the -1 test recommended in coltons other comment works, have 1 expire then try to add a 52 which should succed. This to check an edge case around the max event buffer size.

File: OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp (hunk 6 above)
Comment (Newfloridadad) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: They can cause a crash if you run out of memory, typically MB or KB which I believe the STM is in. If you cap a hashmap/vectors size youll be fine and shave some iterations. Sometimes microcontrolls are finnicy on if they support them at all though. a secondary set/ stack of open event indexes would save you on lookup, not run over size and provide a running count of free event spaces if you did want to stick with an array as the main structure.
All you would have to do is remove values when adding events and add it back when in cleanup(), but realistically youve built a priority queue based on time.

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Are you suggesting making that secondary array dynamic?

Comment (Newfloridadad) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: I wouldnt use a second array but a stack or a set which are inhetintly dynamic, but fill them with all 0 to 'num of event ids' size on creation. They dont grow any bigger then that size so running out of memory wont be an issue and give u O(1) functionality.

Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: On the other hand, opting to pivot to a dynamic data struct removes the need for the second array. I'm a little spooked about undefined behaviour if the event scheduler runs out of room in its array, possible edge case if another class was accidentally adding more events than it needed.

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Ok that makes more sense. Thanks

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: @andrewnash I think that returning an error and dealing with that accordingly is safer than using dynamic memory. STL stuff is rarely used in embedded applications

Comment (andrewnash) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: image

To my knowledge, there is much debate between people far smarter than both of us as to whether STL is appropriate in embedded applications. Also STL != dynamic memory. Regardless of whether you use STL vectors or plain C arrays, you can still use sort(), binary_search(), random_shuffle(), the functions for building and managing heaps, etc these will almost certainly be faster and less buggy than versions you build yourself.... like a linear search for loop

Comment (ShaneTWilliams) [reply OnPod/Node/Common/EventScheduler/src/event_scheduler.cpp]: Fair points 100% but given the fact that I have 4 exams this week I won't be able to make changes to the code anyways until this weekend haha. Let's shelf this until then

**Comment by ShaneTWilliams** [comment]
Changed the main data structure to a map. It's just as fast (as a vector) when iterating and faster in finding available eventId, inserting and erasing.

Changed the `Event` class to a struct because it was only holding data anyways.

Added the tests suggested in the PR.

Moved the whole rig to OnPod/Common from Node/Common.

**Review by kmk142789 — APPROVED** [review]
